mParticle-Apple-SDK/Data Model/MPConsumerInfo.mm

```diff
@@ -307,46 +322,52 @@ - (NSDictionary *)localCookiesDictionary {
 
 #pragma mark Public accessors
 - (NSNumber *)mpId {
-    if (_mpId) {
-        return _mpId;
-    }
-    
-    [self willChangeValueForKey:@"mpId"];
-    
-    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
-    NSString *mpIdString = userDefaults[kMPRemoteConfigMPIDKey];
-    
-    if (mpIdString) {
-        [userDefaults removeMPObjectForKey:kMPRemoteConfigMPIDKey];
-
-        if (sizeof(void *) == 4) { // 32-bit
-            _mpId = [NSNumber numberWithLongLong:(long long)[mpIdString longLongValue]];
-        } else if (sizeof(void *) == 8) { // 64-bit
-            _mpId = [NSNumber numberWithLong:(long)[mpIdString longLongValue]];
+    // Must ensure it's created prior to allowing it to be created
+    dispatch_sync(self.queue, ^{
+        // If we don't have the id, create it.
+        if (!_mpId) {
+            [self willChangeValueForKey:@"mpId"];
+            
+            NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+            NSString *mpIdString = userDefaults[kMPRemoteConfigMPIDKey];
+            
+            if (mpIdString) {
+                [userDefaults removeMPObjectForKey:kMPRemoteConfigMPIDKey];
```

Comment:
Unfortunately, I don't think we can.  The test for nil and the creation of the value all need to be inside of one sync block.  

We could remove the access to NSUserDefaults out of the thread, but that means it would always access NSUserDefaults, even if the  value had already been computed.  That would probably be even less efficient.  

On the other hand, this will usually only happen one time and then it will set the value into _mpId.  After that, it will always have a value and the slow path will be skipped.  

Assume that a background thread is currently inside the block, and the main thread is blocked on the generation of this value.  The main thread will then run once the value has been generated, read the value directly and skip over the slow generation code.  There is the slight overhead of the queue emptying and triggering the next one, but that's pretty small and will only happen once.  Once the value has been generated, everything will just become a test that skips over creating the value which is fast.

Even if there 10 more reads in front of the main queue (unlikely), the reads once the value has been created are going to finish extremely fast since they become a single test instruction that jumps to the return statement.

So yes, there is a overhead but it will only paid once, and no matter which thread it took place on, it still had to be paid first before any of the threads could use the value.  The worst case is when multiple threads create a value and then write to the same address at once since that can result in many weird bugs.  Then you have each thread using it's own UUID for the id, then which one you finally settle on is random, and the pointer could even be corrupted since the code for the assignment statement is not atomic.